src/main/java/io/github/lottetreg/httpserver/Router.java

```diff
@@ -0,0 +1,94 @@
+package io.github.lottetreg.httpserver;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Router {
+  private List<Routable> routes;
+  private Outable out;
+
+  Router(List<Routable> routes) {
+    this.routes = routes;
+    this.out = new Out();
+  }
+
+  Router(List<Routable> routes, Outable out) {
+    this.routes = routes;
+    this.out = out;
+  }
+
+  public HTTPResponse route(HTTPRequest request) {
+    try {
+      Routable route = findMatchingRoute(request);
+      HTTPResponse response = route.getResponse(request);
+
+      if(request.getMethod().equals("OPTIONS")) {
+        response.addHeader("Allow", getAllowedMethods(request));
+      }
+
+      return response;
+
+    } catch (NoMatchingPath e) {
+      this.out.println(e.getMessage());
+      return new HTTPResponse.Builder(404).build();
+
+    } catch (NoMatchingMethodForPath e) {
+      this.out.println(e.getMessage());
+      return new HTTPResponse.Builder(405).build()
+          .addHeader("Allow", getAllowedMethods(request));
+
+    } catch (Exception e) {
+      this.out.println(e.getMessage());
+      return new HTTPResponse.Builder(500).build();
+    }
+  }
+
+  private Routable findMatchingRoute(HTTPRequest request) {
+    String requestPath = request.getURI();
+    String requestMethod = request.getMethod();
+
+    if (noRoutesMatchingPath(requestPath)) {
+      throw new NoMatchingPath(requestPath);
+    }
+
+    Optional<Routable> matchingRoute = routesWithMatchingPath(requestPath)
+        .filter(route -> route.hasMethod(requestMethod))
+        .findFirst();
+
+    if (matchingRoute.isEmpty()) {
+      throw new NoMatchingMethodForPath(requestMethod, requestPath);
+    }
+
```

Comment:
Nice use of Optionals here! You can condense this even further and nix using an `if` statement completely: 
```
return matchingRoute.orElseThrow(() -> new NoMatchingMethodForPath(requestMethod, requestPath);
```

Reply:
OooOOoo I'll add that now! Thanks!

Reply:
Since you asked about Scala the other day ... :) pattern matching on optionals is much nicer compared to java: 
```
matchingRoute match {
  case Some(route) => route // or do whatever with the unwrapped value
  case None => throw someException
} 
```
ok stepping off the soap box..

Reply:
I love pattern matching! After doing minesweeper in Elixir I keep seeing places I wish I could use it in here... Scala++!